vsp.go

```diff
@@ -60,62 +66,298 @@ func (mw *MultiWallet) NewVSPClient(vspHost string, walletID int, purchaseAccoun
-	err := v.w.UnlockWallet(passphrase)
+func (vsp *VSP) PurchaseTickets(ticketCount, expiryBlocks int32, passphrase []byte) error {
+	err := vsp.w.UnlockWallet(passphrase)
 	if err != nil {
 		return translateError(err)
 	}
-	defer v.w.LockWallet()
+	defer vsp.w.LockWallet()
 
 	request := &w.PurchaseTicketsRequest{
 		Count:                int(ticketCount),
-		SourceAccount:        v.purchaseAccount,
-		Expiry:               v.w.GetBestBlock() + expiryBlocks,
+		SourceAccount:        vsp.purchaseAccount,
+		Expiry:               vsp.w.GetBestBlock() + expiryBlocks,
 		MinConf:              DefaultRequiredConfirmations,
-		VSPFeeProcess:        v.PoolFee,
-		VSPFeePaymentProcess: v.ProcessFee,
+		VSPFeeProcess:        vsp.PoolFee,
+		VSPFeePaymentProcess: vsp.ProcessFee,
 	}
 
-	networkBackend, err := v.w.Internal().NetworkBackend()
+	networkBackend, err := vsp.w.Internal().NetworkBackend()
 	if err != nil {
 		return err
 	}
 
-	_, err = v.w.Internal().PurchaseTickets(v.ctx, networkBackend, request)
+	_, err = vsp.w.Internal().PurchaseTickets(vsp.ctx, networkBackend, request)
 	if err != nil {
 		return err
 	}
 
 	return nil
 }
 
+// StartTicketBuyer starts the automatic ticket buyer
+func (vsp *VSP) StartTicketBuyer(balanceToMaintain int64, passphrase []byte) error {
+	if balanceToMaintain < 0 {
+		return errors.New("Negative balance to maintain given")
+	}
+
+	if len(passphrase) > 0 {
+		err := vsp.w.UnlockWallet(passphrase)
+		if err != nil {
+			return translateError(err)
+		}
+	}
+
+	if vsp.mwRef.IsAutoTicketsPurchaseActive(vsp.w.ID) {
+		return errors.New("Ticket buyer already running")
+	}
+
+	go func() {
+		log.Info("Running ticket buyer")
+
+		ctx, cancel := vsp.mwRef.contextWithShutdownCancel()
+		vsp.w.cancelAutoTicketBuyer = cancel
+		err := vsp.ticketBuyer(ctx, balanceToMaintain)
+		if err != nil {
+			log.Errorf("Ticket buyer instance errored: %v", err)
+		}
+
+		vsp.w.cancelAutoTicketBuyer = nil
+	}()
+
+	return nil
+}
+
+func (vsp *VSP) ticketBuyer(ctx context.Context, balanceToMaintain int64) error {
+	c := vsp.w.Internal().NtfnServer.MainTipChangedNotifications()
+	defer c.Done()
+
+	ctx, outerCancel := context.WithCancel(ctx)
+	defer outerCancel()
+	var fatal error
+	var fatalMu sync.Mutex
+
+	var nextIntervalStart, expiry int32
+	var cancels []func()
+	for {
+		select {
+		case <-ctx.Done():
+			defer outerCancel()
+			fatalMu.Lock()
+			err := fatal
+			fatalMu.Unlock()
+			if err != nil {
+				return err
+			}
+			return ctx.Err()
+		case n := <-c.C:
+			if len(n.AttachedBlocks) == 0 {
+				continue
+			}
+
+			tip := n.AttachedBlocks[len(n.AttachedBlocks)-1]
+			w := vsp.w.Internal()
+
+			// Don't perform any actions while transactions are not synced through
+			// the tip block.
+			rp, err := w.RescanPoint(ctx)
+			if err != nil {
+				return err
+			}
+			if rp != nil {
+				log.Debugf("Skipping autobuyer actions: transactions are not synced")
```

Comment:
Is it possible for logs to include the wallet ID/name so users running ticket buyers on multiple wallets can identify which wallet is producing which log message?

Reply:
It is not possible to run ticket buyer on multiple wallets at same time.. you will need to cancel and start a new process with the new wallet.